flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -0,0 +1,268 @@
+	}
+
+	private static ShuffleDescriptor[] getConsumedPartitionShuffleDescriptors(
+			ExecutionEdge[] edges,
+			boolean allowLazyDeployment) throws ExecutionGraphException {
+		ShuffleDescriptor[] shuffleDescriptors = new ShuffleDescriptor[edges.length];
+		// Each edge is connected to a different result partition
+		for (int i = 0; i < edges.length; i++) {
+			shuffleDescriptors[i] =
+				getConsumedPartitionShuffleDescriptor(edges[i], allowLazyDeployment);
+		}
+		return shuffleDescriptors;
+	}
+
+	private static ShuffleDescriptor getConsumedPartitionShuffleDescriptor(
+			ExecutionEdge edge,
+			boolean allowLazyDeployment) throws ExecutionGraphException {
+		IntermediateResultPartition consumedPartition = edge.getSource();
+		Execution producer = consumedPartition.getProducer().getCurrentExecutionAttempt();
+
+		ExecutionState producerState = producer.getState();
+		Optional<ResultPartitionDeploymentDescriptor> consumedPartitionDescriptor =
+			producer.getProducedPartition(consumedPartition.getPartitionId());
+
+		ResultPartitionID consumedPartitionId = new ResultPartitionID(
+			consumedPartition.getPartitionId(),
+			producer.getAttemptId());
+
+		return getConsumedPartitionShuffleDescriptor(
+			consumedPartitionId,
+			consumedPartition.getResultType(),
+			consumedPartition.isConsumable(),
+			producerState,
+			allowLazyDeployment,
+			consumedPartitionDescriptor.orElse(null));
+	}
+
+	@VisibleForTesting
+	static ShuffleDescriptor getConsumedPartitionShuffleDescriptor(
+			ResultPartitionID consumedPartitionId,
+			ResultPartitionType resultPartitionType,
+			boolean isConsumable,
+			ExecutionState producerState,
+			boolean allowLazyDeployment,
+			@Nullable ResultPartitionDeploymentDescriptor consumedPartitionDescriptor) throws ExecutionGraphException {
+		// The producing task needs to be RUNNING or already FINISHED
+		if ((resultPartitionType.isPipelined() || isConsumable) &&
+			consumedPartitionDescriptor != null &&
+			isProducerAvailable(producerState)) {
+			// partition is already registered
+			return consumedPartitionDescriptor.getShuffleDescriptor();
+		}
+		else if (allowLazyDeployment) {
+			// The producing task might not have registered the partition yet
+			return new UnknownShuffleDescriptor(consumedPartitionId);
+		}
+		else {
+			// throw respective exceptions
+			handleConsumedPartitionShuffleDescriptorErrors(
+				consumedPartitionId,
+				resultPartitionType,
+				isConsumable,
+				producerState);
+			return null;
+		}
+	}
+
+	private static void handleConsumedPartitionShuffleDescriptorErrors(
+			ResultPartitionID consumedPartitionId,
+			ResultPartitionType resultPartitionType,
+			boolean isConsumable,
+			ExecutionState producerState) throws ExecutionGraphException {
+		if (isProducerFailedOrCanceled(producerState)) {
+			String msg = "Trying to schedule a task whose inputs were canceled or failed. " +
+				"The producer is in state " + producerState + ".";
+			throw new ExecutionGraphException(msg);
```

Comment:
Should we keep the previous `IllegalStateException` here during refactoring, also for the following else?